### PR TITLE
feat(lsp): ts language service scopes

### LINKED ISSUE #24293: Multi deno.json resolver scopes fails for JSX
Version: Deno 1.44.4

From https://github.com/denoland/deno/pull/24206

I have a web application that renders HTML with hono/jsx. However I want to create some aliases for my JS client to use React. 

This is my project structure:

image

deno.json (server)

```json
{
  "lock": false,
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "hono/jsx"
  },
  "imports": {
    "@unocss/core": "https://esm.sh/@unocss/core@0.58.0",
    "@unocss/preset-uno": "https://esm.sh/@unocss/preset-uno@0.58.0",
    "@unocss/reset": "https://esm.sh/@unocss/reset@0.58.0/tailwind-compat.css",
    "hono": "https://deno.land/x/hono@v3.11.7/mod.ts",
    "hono/html": "https://deno.land/x/hono@v3.11.7/helper/html/index.ts",
    "hono/jsx/jsx-runtime": "https://deno.land/x/hono@v3.11.7/jsx/jsx-runtime.ts",
    "$std/": "https://deno.land/std@0.209.0/"
  }
}

```

Notice how client.js is in red, this is because it can't resolve `react` which is expected:

image

So I add a deno.json on the client:

```json
{
  "lock": false,
  "imports": {
    "react": "https://esm.sh/react@18.3.1"
  }
}

```

Now client.js is in green:

image

But Layout.tsx now fails because is loading React JSX by default but hono/jsx:

image

Is there a way to apply JSX configuration only to the directory that holds deno.json on the client? Notice Layout.tsx is a folder upper static folder; so it should read the configuration of deno.json on the server.

image

Thanks.

### PR DESCRIPTION
**Pull request by nayeemrmn**
Closes https://github.com/denoland/vscode_deno/issues/488.
Closes #24293.

### CODE REVIEW

**Review by marvinhagemeister — APPROVED** [review]

File: cli/lsp/documents.rs [hunk 1]
```diff
@@ -926,9 +947,9 @@ pub struct Documents {
   /// The npm package requirements found in npm specifiers.
   npm_reqs_by_scope:
     Arc<BTreeMap<Option<ModuleSpecifier>, BTreeSet<PackageReq>>>,
-  /// Gets if any document had a node: specifier such that a @types/node package
-  /// should be injected.
-  has_injected_types_node_package: bool,
+  /// Config scopes that contain a node: specifier such that a @types/node
+  /// package should be injected.
+  scopes_with_node_specifier: Arc<HashSet<Option<ModuleSpecifier>>>,
```
Comment (dsherret) [inline cli/lsp/documents.rs:952]: It's a little strange to see a hashmap containing an Option. Maybe we should bury this behaviour within a struct and make the public api of the struct better explain what's going on here?

Comment (dsherret) [reply cli/lsp/documents.rs:952]: Oh, nvm, I see what this means. We may want to give `Option<ModuleSpecifier>` a name like `ScopeKey` or something?

File: cli/lsp/tsc.rs [hunk 2]
```diff
@@ -817,35 +925,60 @@ impl TsServer {
...
+      let Some(mut locations) = locations else {
+        continue;
+      };
+      for location in &mut locations {
+        location.normalize(&self.specifier_map)?;
+      }
+      all_locations.extend(locations);
+    }
+    if all_locations.is_empty() {
+      return Ok(None);
```
Comment (dsherret) [inline cli/lsp/tsc.rs:967]: I wonder why this returns an option. Is something relying on this returning `None`? Maybe we can just return a vec now? If not, it looks like the behaviour might have changed?

Comment (nayeemrmn) [reply cli/lsp/tsc.rs:967]: It's just preserving the TS API (https://github.com/denoland/deno/blob/v1.37.1/cli/tsc/dts/typescript.d.ts#L6228), which probably tries to match the LSP API. With all of our intermediate plumbing it's basically meaningless.

**Review by dsherret — APPROVED** [review]
Amazing!

**Comment by sant123** [comment]
Thank you so much 🙏🏼

**Pull request merged**